File: matlab/orth/orth_hh_lvl1.m
%
   function [ q, t, r, a ] = orth_hh_lvl1( V, tau, a ) 
%  
      m = size(V,1);
      j = size(V,2)+1;
%
      r = zeros(j,1);
      q = zeros(m,1);
%
      for i = 1:j-1,
         alpha = tau(i) * ( V(i:m,i)' * a(i:m,1) );       
         a(i:m,1) = a(i:m,1) - V(i:m,i) * alpha;          
      end 
%
      if (j>1), r(1:j-1,1) = a(1:j-1,1); end;           
%
      normx = norm( a(j+1:m,1), 2);                       
      norma = sqrt( normx*normx +  a(j,1)*a(j,1) );
      if ( a(j,1) > 0.0e+00 ) r(j,1) = - norma; else r(j,1) = norma; end
      if ( a(j,1) > 0.0e+00 ) a(j,1) = a(j,1) + norma; else a(j,1) = a(j,1) - norma; end
      a(j+1:m,1) = a(j+1:m,1) / a(j,1);
      t = 2.0e+00 / ( 1.0e+00 + ( normx / a(j,1) )^2 );
      a(1:j-1,1) = 0.0e+00;
      a(j,1) = 1.0e+00;
%
      q(j,1) = - t;
      q(j+1:m,1) = a(j+1:m,1) * q(j,1);
      q(j,1) = 1.0e+00 + q(j,1);
      for i = j-1:-1:1,
        q(i,1) = - tau(i) * ( V(i+1:m,i)' * q(i+1:m,1) );  
        q(i+1:m,1) = q(i+1:m,1) + V(i+1:m,i) * q(i,1);     
      end
%
   end

File: julia/orth/orth_hh_lvl1.jl
using LinearAlgebra

function orth_hh_lvl1(V::AbstractMatrix, tau::AbstractVector, a::AbstractVector)
    m = size(V, 1)
    j = size(V, 2) + 1
    r = zeros(j)
    q = zeros(m)
    a = copy(a)

    for i in 1:j-1
        alpha = tau[i] * (V[i:m, i]' * a[i:m])
        a[i:m] .-= V[i:m, i] .* alpha
    end

    if j > 1
        r[1:j-1] = a[1:j-1]
    end

    normx = norm(a[j+1:m])
    norma = sqrt(normx^2 + a[j]^2)
    r[j] = a[j] > 0.0 ? -norma : norma
    a[j] = a[j] > 0.0 ? a[j] + norma : a[j] - norma
    a[j+1:m] ./= a[j]
    t = 2.0 / (1.0 + (normx / a[j])^2)
    a[1:j-1] .= 0.0
    a[j] = 1.0

    q[j] = -t
    q[j+1:m] = a[j+1:m] .* q[j]
    q[j] = 1.0 + q[j]
    for i in j-1:-1:1
        q[i] = -tau[i] * (V[i+1:m, i]' * q[i+1:m])
        q[i+1:m] .+= V[i+1:m, i] .* q[i]
    end

    return q, t, r, a
end
